Implement emit-h

tetsuo-cpp opened the pull request:
Closes #6533.

tetsuo-cpp commented on src/codegen/c.zig:
```diff
@@ -19,24 +21,31 @@ fn map(allocator: *std.mem.Allocator, name: []const u8) ![]const u8 {
                 try writer.writeAll("size_t");
             } else {
-                return ctx.file.fail(ctx.decl.src(), "TODO implement int types", .{});
+                return C.failImpl(allocator, error_msg, decl.src(), "TODO implement int types", .{});
```
Not sure whether this is the right way to do this. I wanted to make this independent of `Context` but unfortunately, it's caused a bunch of ugly parameters to get piped through these helpers.

pixelherodev replied:
If you need to, you could add a `ctx.fail` function which handles it, but we need to decide how errors should be handled here before worrying about the implementation anyways.

tetsuo-cpp replied:
My bad, I misunderstood where `Context` is coming from. I'll revert this portion of the change.

tetsuo-cpp commented on src/link/C.zig:
```diff
@@ -13,15 +13,29 @@ const C = @This();
 
 pub const base_tag: File.Tag = .c;
 
+pub const HeaderInfo = struct {
```
The header specific stuff uses this instead of a `link.C`. In the normal codegen case, we pass in `link.C.header_info` and when we're generating C headers, we create our own `HeaderInfo`.

pixelherodev commented:
I'll review this tonight or tomorrow before getting back to work on the CBE :)

pixelherodev commented:
Alright, reviewing now! :)

@tetsuo-cpp Are you on IRC?

pixelherodev commented on src/codegen/c.zig:
```diff
@@ -51,13 +60,19 @@ fn renderValue(ctx: *Context, writer: std.ArrayList(u8).Writer, T: Type, val: Va
 }
 
-fn renderFunctionSignature(ctx: *Context, writer: std.ArrayList(u8).Writer, decl: *Decl) !void {
+fn renderFunctionSignature(
```
Same as renderType; stick to using the Context.

pixelherodev commented on src/codegen/c.zig:
```diff
@@ -83,6 +98,28 @@ pub fn generate(file: *C, decl: *Decl) !void {
+            var error_msg: *ErrorMsg = undefined;
+            const writer = header_info.buf.writer();
+            renderFunctionSignature(arena, &error_msg, header_info, writer, decl) catch |err| {
+                if (err == error.AnalysisFail) {
```
This also means that the error_msg is never used, which means you can probably remove it as a parameter and all related logic.

tetsuo-cpp replied:
Yeah, I was only doing this because TODOs return `error.AnalysisFail`. If we're saying that we can just replace those with panics (which makes sense, now that I think about it), we can get rid of this.

pixelherodev replied:
On second thought, leave it alone. Normal codegen uses AnalysisFail for TODOs; we should follow it. This should be fine, ignore my initial comments.

pixelherodev commented on src/link/C.zig:
```diff
@@ -53,13 +67,23 @@ pub fn openPath(allocator: *Allocator, sub_path: []const u8, options: link.Optio
+    args: anytype,
+) error{ AnalysisFail, OutOfMemory } {
+    error_msg.* = try Compilation.ErrorMsg.create(allocator, src, format, args);
     return error.AnalysisFail;
```
Also, what was the reasoning for this? Why are you using a two-level pointer like this? This looks wrong.

tetsuo-cpp replied:
The design seems to be that the Context's `error_msg` is a pointer that gets written to with allocated memory whenever there is an error. So if I'm writing to the pointer, I think it has to be `**`. Is that not right?

pixelherodev commented:
@tetsuo-cpp Do you want to fix this up, or do you want me to? I'm happy either way :)

pixelherodev commented:
Ah, right, one other thing: you need to add tests. However, there's no test harness for headers yet. I was planning on working on tests today anyways, I can add the harness and tests in a separate PR if you want to do the rest of the work and leave the tests to me. Alternately, if you want, I can (as mentioned earlier) take over this PR and do the tests myself, or I can help you with the full PR + tests and leave the work to you on this one.

This is entirely up to you, just let me know how you want to proceed. I'm going to work on unrelated CBE test integration (running the tests with `zig cc` now that we have stage1+stage2 integration :) in the meanwhile, so no rush.

tetsuo-cpp commented:
Thanks for taking a look @pixelherodev.

> Are you on IRC?

I don't use IRC at the moment. I can get set up if it's helpful.
Alternatively, I'm on the Zig Discord under "asc" if you use that.

> Do you want to fix this up, or do you want me to? I'm happy either way :)

I'm happy to continue working on this until it's mergeable. Most of these comments look straightforward so I'll address them shortly.

> Ah, right, one other thing: you need to add tests. However, there's no test harness for headers yet. I was planning on working on tests today anyways, I can add the harness and tests in a separate PR if you want to do the rest of the work and leave the tests to me.

Sure, let's do that. Once you've written the harness, I'm happy to help with writing tests if you'd like.

tetsuo-cpp commented on src/Compilation.zig:
```diff
@@ -1252,22 +1276,44 @@ pub fn performAllTheWork(self: *Compilation) error{ TimerUnsupported, OutOfMemor
-                        ));
-                        decl.analysis = .codegen_failure_retryable;
-                    },
+                self.bin_file.updateDecl(module, decl) catch |err| {
```
I added a `return` after catching an error. My thinking is that there's no point in doing the header generation if the code generation for that decl didn't work.

tetsuo-cpp commented:
@pixelherodev 

I've addressed your comments and marked the ones that aren't relevant anymore as "resolved".

Since we decided to continue returning `error.AnalysisFail` in the TODO case, most of the error handling looks the same.

The main sticking point this time was reverting back to using the `Context` object as an argument. Some of the stuff in `Context` (mainly the `file`) aren't relevant in the header generation path. What I've done is marked this member as an optional and in the header path, assigned it to `null`. It seems a bit fragile but the idea is that in any of the logic used in header generation, we shouldn't be touching `file`.

Another alternative would be to rip `file` out of the `Context` and just pass it in a separate argument. That way, anything that is used in the header generation path won't take a `file` argument and we can't make this mistake of trying to use `file` when it's `null`. I realize this adds an additional function argument but it seems better to me. What do you think?

Since `Context` is just an internal thing and `link.C` interacts with code generation via the `file`, I need some way of propagating error messages to `file`. But if we get an error in the header generation logic, then I can't rely on calling `ctx.file.fail` since in the `emit-h` case, it will be `null`. So I've introduced `ctx.fail` and then added some logic at the top level to transfer the `error_msg` across to the `file` so it can be used by the caller. That part is at `codegen/c.zig:186`.

Sorry if this isn't what you meant, I'm still finding my feet here.

pixelherodev commented:
`file` as a separate argument is fine IMO - that's entirely up to you.

I think this is good, just needs the testing work - I'll try to have that ready for you tonight :)

pixelherodev commented:
Mind if I open a PR against your branch splitting this into multiple commits?

tetsuo-cpp commented:
> Mind if I open a PR against your branch splitting this into multiple commits?

Sure, no problem.

pixelherodev commented:
Okay, I think I should have a PR for header tests shortly :)

pixelherodev commented:
@tetsuo-cpp  https://github.com/tetsuo-cpp/zig/pull/2/

pixelherodev commented:
Last problem: emit-h in compilation and c_header shouldn't coexist, and should be merged.

pixelherodev commented:
Actually, this is also missing emit_h enablement in `src/main.zig`

tetsuo-cpp commented:
> Last problem: emit-h in compilation and c_header shouldn't coexist, and should be merged.

Done. I added an `?EmitLoc` member to `c_header`.

> Actually, this is also missing emit_h enablement in `src/main.zig`

Done. Thanks, I missed that. Am I understanding correctly that if we do `build-obj` or `build-lib`, we're supposed to automatically emit a header even without `femit-h`?

> @tetsuo-cpp [tetsuo-cpp#2](https://github.com/tetsuo-cpp/zig/pull/2)

Thanks for doing that. It looks good to me.

How do we want to do this? I'm thinking that I should merge that PR into my branch now and then I'll push a follow up commit adding some unit tests for the header generation. Do you agree with that?

pixelherodev commented:
> . Am I understanding correctly that if we do build-obj or build-lib, we're supposed to automatically emit a header even without femit-h?

> (default) Do not generate a C header file (.h)

Nope!

> I'm thinking that I should merge that PR into my branch now and then I'll push a follow up commit adding some unit tests for the header generation. Do you agree with that?

Yep!

Once that's done, we can bump this up to the core team for further review and merging :)

tetsuo-cpp commented:
> > (default) Do not generate a C header file (.h)
> 
> Nope!

The changes that I uncommented in `main.zig` look like they're doing that. The argument parsing was already working, that chunk looks like it's enabling it by default in certain scenarios.

> Once that's done, we can bump this up to the core team for further review and merging :)

Sounds good! I'll prepare some tests over the weekend.

pixelherodev commented:
Ah, I see.

I'd say either leave it as is and change the docs (`zig build-* --help` output), or make it always off without an explicit flag, at your discretion. As long as docs and behavior match, I don't think it's too important - and if it is, it can be changed trivially later instead of bikeshedding it now.

tetsuo-cpp commented:
@pixelherodev 
Ok, I've added some tests now.

The only other changes since last time are that I've disabled header generation unless `-femit-h` is supplied and I made a minor change to the test code to avoid running with header generation for non-CBE tests since some of them use things that aren't supported yet like `u64`.

uael commented on src/Compilation.zig:
```diff
@@ -1138,6 +1141,20 @@ pub fn update(self: *Compilation) !void {
         }
     }
+
+    // If we've chosen to emit a C header, flush the header the disk.
```
```suggestion
    // If we've chosen to emit a C header, flush the header to the disk.
```

tetsuo-cpp replied:
Rebased as part of b323209.

pixelherodev commented:
Okay,if you merge the suggested fix for the comment, I think this has my 
approval :)

tetsuo-cpp commented:
> Okay,if you merge the suggested fix for the comment, I think this has my approval :)

Awesome. Done!

phifty commented:
Are there any news on this?

tetsuo-cpp commented:
> Are there any news on this?

This patch is ready to go from my point of view. Just pending a review + merge from the team.

phifty commented:
Ok thanks! I'm looking forward to this. It's preventing me to update my zig version.

Vexu commented on src/test.zig:
```diff
@@ -611,6 +636,22 @@ pub const TestContext = struct {
+                    defer file.close();
+                    var out = file.reader().readAllAlloc(arena, 1024 * 1024) catch @panic("Unable to read headeroutput!");
+
+                    if (expected_output.len != out.len) {
+                        std.debug.print("\nTransformed header length differs:\n================\nExpected:\n================\n{}\n================\nFound:\n================\n{}\n================\nTest failed.\n", .{ expected_output, out });
+                        std.process.exit(1);
+                    }
+                    for (expected_output) |e, i| {
+                        if (out[i] != e) {
+                            std.debug.print("\nTransformed header differs:\n================\nExpected:\n================\n{}\n================\nFound:\n================\n{}\n================\nTest failed.\n", .{ expected_output, out });
+                            std.process.exit(1);
+                        }
+                    }
```
These should use `std.testing.expectEqualStrings`.

tetsuo-cpp replied:
As @pixelherodev mentioned, it's done like this elsewhere in `test.zig`. I'm planning to follow up with a PR after this to make this change across the file.

Vexu commented on src/Compilation.zig:
```diff
@@ -2828,7 +2870,7 @@ fn updateStage1Module(comp: *Compilation, main_progress_node: *std.Progress.Node
     man.hash.add(comp.bin_file.options.is_test);
     man.hash.add(comp.bin_file.options.emit != null);
-    man.hash.addOptionalEmitLoc(comp.emit_h);
+    man.hash.add(comp.c_header != null);
```
There should be a `addOptionalEmitLoc(header.emit_loc.?)` here if `c_header != null`.

tetsuo-cpp replied:
e5fb9d0

Vexu commented on src/codegen/c.zig:
```diff
@@ -158,18 +198,18 @@ fn genFn(file: *C, decl: *Decl) !void {
-                .call => try genCall(&ctx, inst.castTag(.call).?),
-                .add => try genBinOp(&ctx, inst.cast(Inst.BinOp).?, "+"),
-                .sub => try genBinOp(&ctx, inst.cast(Inst.BinOp).?, "-"),
+                .assembly => try genAsm(&ctx, file, inst.castTag(.assembly).?),
```
Why remove `file` from the context if you're just going to pass it to nearly everything anyways?

tetsuo-cpp replied:
The `file` contains everything needed to output C source when the CBE is selected (buffers for main, constants, etc). That doesn't apply when we're generating a header and we may not even have one since we could be using a different backend in combination with `emit-h`.

Therefore, I've removed it from the context and passed it as a separate argument in the non-header code paths. Another idea could be to make it optional within the context but that seemed more error-prone to me.

pixelherodev commented:
File should stay in the context.

Re testing expect vs print and exit, I think I copied that style from the 
surrounding code - might be better to leave it alone here and do a full 
cleanup in a follow up PR.

tetsuo-cpp commented:
@Vexu
Thanks for the review. I've just rebased to fix the merge conflict.
